Register ATen ops with c10

smessmer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #26128 [wip] custom op API improvements
* **#23667 Register ATen ops with c10**
* #25947 Allow overwriting catch-all kernels
* #26114 Make schema part of RegisterOperators::Options

Changes in this PR:
- For each operator with use_c10_dispatcher: True, additionally generate a c10 registration line in TypeDefault.cpp, CPUType.cpp, and other backend files.
- This doesn't change globalATenDispatch yet, the c10 registration is purely additional and the operator calling path doesn't change. A diff further up the stack will change these things.
- Enable the use_c10_dispatcher: True flag for about ~70% of operators
This also changes the c10->jit operator export because ATen ops are already exported to JIT directly and we don't want to export the registered c10 ops because they would clash
- For this, we need a way to recognize if a certain operator is already moved from ATen to c10, this is done by generating a OpsAlreadyMovedToC10.cpp file with the list. A diff further up in the stack will also need this file to make sure we don't break the backend extension API for these ops.

Reasons for some ops to be excluded (i.e. not have the use_c10_dispatcher flag set to true):
- Tensor?(a!) (i.e. optional tensor with annotations) not supported in c++ function schema parser yet
- -> void in native_functions.yaml vs -> () expected by function schema parser
- out functions have different argument order in C++ as in the jit schema
Tensor? (i.e. optional tensor) doesn't work nicely with undefined tensor sometimes being undefined tensor and sometimes being None.
- fixed-size arrays like int[3] not supported in c10 yet
These will be fixed in separate diffs and then the exclusion tag will be removed.

Differential Revision: [D16603131](https://our.internmc.facebook.com/intern/diff/D16603131/)

gchanan commented:
what's the benefit of the registration?  Are you going to try to move over the eager ops ? is it so the JIT can universally use the c10 dispatcher? etc.

when are you planning to address the above limitations?

gchanan commented:
what happens if someone adds a function with one of the above limitations?  Do they get a compiler error?  An error if they try to dispatch to the function (via JIT?)?

There is no documentation in the README.

Why was the choice made to exclude rather than include via the flag?

smessmer commented:
Yes, I'm planning to move over all of native_functions.yaml to c10 so that we have one dispatch mechanism for everything instead of multiple ones.

Most of the above limitations will be addressed very soon. The argument order for out functions probably needs a bit more discussion with you on how to do it first.

Adding a function that violates the limitations is a compiler error.

The choice to exclude instead of include is that over time the code base can PR by PR get better and better with fewer uses of the flag, until the flag is entirely gone.

I will add a section to the README

gchanan commented:
that sounds generally reasonable.  Do I understand correctly: if a user adds a function that violates the limitations but doesn't put the exclude, they will get a compiler error?  If so, we shouldn't do that.  We should either detect the error and print a message or have the annotation be opt-in.

li-roy commented on aten/src/ATen/function_wrapper.py:
```diff
@@ -1158,6 +1173,9 @@ def add_namedtensor_enabled_macro(code):
                 check_namedtensor_enabled(NATIVE_DISPATCH_DEFINITION_DEFAULT.substitute(option)))
             top_env['function_registrations'].append(
                 check_namedtensor_enabled(DEFAULT_FUNCTION_REGISTRATION.substitute(option)))
+            if not option['exclude_from_c10_dispatcher']:
```
My main concern with this is that if we have different registrations, it will make backend extensions very confusing. I guess this patch itself seems fine, but once if we start calling through c10 path, it will be confusing.

smessmer replied:
Do you think we can have globalATenDispatch forward registration calls to c10 as a temporary solution? This way, we only have to switch backends once when all ops are on c10.

li-roy replied:
Yeah I think that's reasonable, especially if we're already generating a list of exclusions for the jit registry, we can just use it for this purpose as well.

li-roy commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -105,13 +112,15 @@
   named_guard: False
 
 - func: abs_(Tensor(a!) self) -> Tensor(a!)
+  exclude_from_c10_dispatcher: True
```
I think you forgot to mention in the description that inplace and _out ops can't be registered at the moment because of Tensor & in their signature. Did you decide on an approach for this yet btw?

smessmer replied:
no. out functions don't work anyhow because jit and c++ have a different argument order. Not sure yet about how to fix this.

dzhulgakov replied:
according to @gchanan, the reason for is that _out variants of functions have `out` argument as `kwarg_only` in python, but in C++ it's logical to always provide it while still have other args with default values. Thus the only option is to make it the first argument

Relevant code seems to be:
https://github.com/pytorch/pytorch/blob/33a1c30cb117c2e39126e49b5ec4a2b9b7a64b0d/aten/src/ATen/native_parse.py#L190
https://github.com/pytorch/pytorch/blob/33a1c30cb117c2e39126e49b5ec4a2b9b7a64b0d/tools/jit/gen_jit_dispatch.py#L452

Note that all C++ APIs today: public API, kernel implementations all use "reordered" variant with `out` argument first. TorchScript parsing and the Stack in JIT uses the "original" order.

If we try to introduce reordering logic in c10 it'd cause a mess when switching between boxed and unboxed APIs.

I see several options:
1) switch kernel definitions to "original" order (giant codemod) and keep reordering only in generated C++ API exposed to user
2) try mirroring reordering logic in c10 such that all boxing/unboxing functions do reordering
3) if we give up the idea of "no codegen" (which I think is reasonable in mid-term). Then for native_functions we can provide ability to codegen boxing/unboxing wrappers and just register them with c10 as hooks (and utilize current codegen). Default hook would be the regular templated wrappers generated automatically based on C++ signature for cases where there's no reordering.

gchanan replied:
I'd view this problem as analogous to the TensorOptions issue.  There, we do some codegen in order to make the C++ API nice (good!), but then need to reimplement the crazy logic in multiple places (bad!).  All of these endpoints don't need to use the fancy C++ signature, they aren't user endpoints, a dumb signature is totally fine.

Now, you may or may not want a _single_ way of telling downstream components that they should actually map to a different C++ function call so we could have private implementation calls (which would allow us to e.g. only expose a single "public" function for factories and out functions, as we do now), but that logic can be super simple.

smessmer commented:
@gchanan I changed it to be an op-in flag instead of opt-out

dzhulgakov commented:
I think you need to patch some more tests - RPC now uses overload signatures too, it seems

gchanan commented on aten/src/ATen/native/README.md:
````diff
@@ -299,6 +299,18 @@ with other components of PyTorch in order to reduce overall complexity.
+use_c10_dispatcher: True
+```
+
+This will indicate that the func signature only uses features supported by
````
It's not that important that people modifying native_functions.yaml undersatnd your plan for the c10 dispatcher, but it is important that they understand what to do for their change.  So you should add some guidance like:
1) mention the current limitations
2) suggest if the func doesn't have those limitations, to set the flag
3) mention if everything compiles, that it should be fine to use dispatcher, otherwise turn it off

Something like that.

smessmer replied:
updated it

ezyang replied:
I don't see any guidance about current limitations in the doc here?

gchanan commented on aten/src/ATen/function_wrapper.py:
```diff
@@ -119,6 +119,18 @@ def TypedDict(name, attrs, total=True):  # type: ignore
 .registerOp<${return_type} (${formals_types})>(Backend::${Backend}, "${schema_string}", &${Type}::${api_name})
```
when I look at the generated code for this in TypeDefault I see that the backend here is Undefined?  Why?

li-roy replied:
The TypeDefault functions are stored in the Undefined entry of the array, just because. No specific reason for it.

smessmer committed:
Update on "Register ATen ops with c10"


Reasons for some ops to be excluded:
- `Tensor?(a!)` (i.e. optional tensor with annotations) not supported in c++ function schema parser yet
- `-> void` in native_functions.yaml vs `-> ()` expected by function schema parser
- out functions have different argument order in C++ as in the jit schema
- `Tensor?` (i.e. optional tensor) doesn't work nicely with undefined tensor sometimes being undefined tensor and sometimes being None.
- fixed-size arrays like `int[3]` not supported in c10 yet

These will be fixed in separate diffs and then the exclusion tag will be removed.

Differential Revision: [D16603131](https://our.internmc.facebook.com/intern/diff/D16603131/)

smessmer committed:
Update on "Register ATen ops with c10"


Reasons for some ops to be excluded:
- `Tensor?(a!)` (i.e. optional tensor with annotations) not supported in c++ function schema parser yet
- `-> void` in native_functions.yaml vs `-> ()` expected by function schema parser
- out functions have different argument order in C++ as in the jit schema
- `Tensor?` (i.e. optional tensor) doesn't work nicely with undefined tensor sometimes being undefined tensor and sometimes being None.
- fixed-size arrays like `int[3]` not supported in c10 yet

These will be fixed in separate diffs and then the exclusion tag will be removed.

Differential Revision: [D16603131](https://our.internmc.facebook.com/intern/diff/D16603131/)

smessmer committed:
Update on "Register ATen ops with c10"


Reasons for some ops to be excluded:
- `Tensor?(a!)` (i.e. optional tensor with annotations) not supported in c++ function schema parser yet
- `-> void` in native_functions.yaml vs `-> ()` expected by function schema parser
- out functions have different argument order in C++ as in the jit schema
- `Tensor?` (i.e. optional tensor) doesn't work nicely with undefined tensor sometimes being undefined tensor and sometimes being None.
- fixed-size arrays like `int[3]` not supported in c10 yet

These will be fixed in separate diffs and then the exclusion tag will be removed.

Differential Revision: [D16603131](https://our.internmc.facebook.com/intern/diff/D16603131/)

smessmer committed:
Update on "Register ATen ops with c10"


Reasons for some ops to be excluded:
- `Tensor?(a!)` (i.e. optional tensor with annotations) not supported in c++ function schema parser yet
- `-> void` in native_functions.yaml vs `-> ()` expected by function schema parser
- out functions have different argument order in C++ as in the jit schema
- `Tensor?` (i.e. optional tensor) doesn't work nicely with undefined tensor sometimes being undefined tensor and sometimes being None.
- fixed-size arrays like `int[3]` not supported in c10 yet

These will be fixed in separate diffs and then the exclusion tag will be removed.

Differential Revision: [D16603131](https://our.internmc.facebook.com/intern/diff/D16603131/)

smessmer committed:
Update on "Register ATen ops with c10"


Reasons for some ops to be excluded:
- `Tensor?(a!)` (i.e. optional tensor with annotations) not supported in c++ function schema parser yet
- `-> void` in native_functions.yaml vs `-> ()` expected by function schema parser
- out functions have different argument order in C++ as in the jit schema
- `Tensor?` (i.e. optional tensor) doesn't work nicely with undefined tensor sometimes being undefined tensor and sometimes being None.
- fixed-size arrays like `int[3]` not supported in c10 yet

These will be fixed in separate diffs and then the exclusion tag will be removed.

Differential Revision: [D16603131](https://our.internmc.facebook.com/intern/diff/D16603131/)

smessmer committed:
Update on "Register ATen ops with c10"


Reasons for some ops to be excluded:
- `Tensor?(a!)` (i.e. optional tensor with annotations) not supported in c++ function schema parser yet
- `-> void` in native_functions.yaml vs `-> ()` expected by function schema parser
- out functions have different argument order in C++ as in the jit schema
- `Tensor?` (i.e. optional tensor) doesn't work nicely with undefined tensor sometimes being undefined tensor and sometimes being None.
- fixed-size arrays like `int[3]` not supported in c10 yet

These will be fixed in separate diffs and then the exclusion tag will be removed.

Differential Revision: [D16603131](https://our.internmc.facebook.com/intern/diff/D16603131/)

smessmer commented:
@gchanan I updated the README

smessmer committed:
Update on "Register ATen ops with c10"


Reasons for some ops to be excluded:
- `Tensor?(a!)` (i.e. optional tensor with annotations) not supported in c++ function schema parser yet
- `-> void` in native_functions.yaml vs `-> ()` expected by function schema parser
- out functions have different argument order in C++ as in the jit schema
- `Tensor?` (i.e. optional tensor) doesn't work nicely with undefined tensor sometimes being undefined tensor and sometimes being None.
- fixed-size arrays like `int[3]` not supported in c10 yet

These will be fixed in separate diffs and then the exclusion tag will be removed.

Differential Revision: [D16603131](https://our.internmc.facebook.com/intern/diff/D16603131/)

smessmer committed:
Update on "Register ATen ops with c10"


Reasons for some ops to be excluded:
- `Tensor?(a!)` (i.e. optional tensor with annotations) not supported in c++ function schema parser yet
- `-> void` in native_functions.yaml vs `-> ()` expected by function schema parser
- out functions have different argument order in C++ as in the jit schema
- `Tensor?` (i.e. optional tensor) doesn't work nicely with undefined tensor sometimes being undefined tensor and sometimes being None.
- fixed-size arrays like `int[3]` not supported in c10 yet

These will be fixed in separate diffs and then the exclusion tag will be removed.

Differential Revision: [D16603131](https://our.internmc.facebook.com/intern/diff/D16603131/)

smessmer committed:
Update on "Register ATen ops with c10"


Reasons for some ops to be excluded:
- `Tensor?(a!)` (i.e. optional tensor with annotations) not supported in c++ function schema parser yet
- `-> void` in native_functions.yaml vs `-> ()` expected by function schema parser
- out functions have different argument order in C++ as in the jit schema
- `Tensor?` (i.e. optional tensor) doesn't work nicely with undefined tensor sometimes being undefined tensor and sometimes being None.
- fixed-size arrays like `int[3]` not supported in c10 yet

These will be fixed in separate diffs and then the exclusion tag will be removed.

Differential Revision: [D16603131](https://our.internmc.facebook.com/intern/diff/D16603131/)

smessmer committed:
Update on "Register ATen ops with c10"


Reasons for some ops to be excluded:
- `Tensor?(a!)` (i.e. optional tensor with annotations) not supported in c++ function schema parser yet
- `-> void` in native_functions.yaml vs `-> ()` expected by function schema parser
- out functions have different argument order in C++ as in the jit schema
- `Tensor?` (i.e. optional tensor) doesn't work nicely with undefined tensor sometimes being undefined tensor and sometimes being None.
- fixed-size arrays like `int[3]` not supported in c10 yet

These will be fixed in separate diffs and then the exclusion tag will be removed.

Differential Revision: [D16603131](https://our.internmc.facebook.com/intern/diff/D16603131/)

smessmer committed:
Update on "Register ATen ops with c10"


Reasons for some ops to be excluded:
- `Tensor?(a!)` (i.e. optional tensor with annotations) not supported in c++ function schema parser yet
- `-> void` in native_functions.yaml vs `-> ()` expected by function schema parser
- out functions have different argument order in C++ as in the jit schema
- `Tensor?` (i.e. optional tensor) doesn't work nicely with undefined tensor sometimes being undefined tensor and sometimes being None.
- fixed-size arrays like `int[3]` not supported in c10 yet

These will be fixed in separate diffs and then the exclusion tag will be removed.

Differential Revision: [D16603131](https://our.internmc.facebook.com/intern/diff/D16603131/)

smessmer committed:
Update on "Register ATen ops with c10"


Reasons for some ops to be excluded:
- `Tensor?(a!)` (i.e. optional tensor with annotations) not supported in c++ function schema parser yet
- `-> void` in native_functions.yaml vs `-> ()` expected by function schema parser
- out functions have different argument order in C++ as in the jit schema
- `Tensor?` (i.e. optional tensor) doesn't work nicely with undefined tensor sometimes being undefined tensor and sometimes being None.
- fixed-size arrays like `int[3]` not supported in c10 yet

These will be fixed in separate diffs and then the exclusion tag will be removed.

Differential Revision: [D16603131](https://our.internmc.facebook.com/intern/diff/D16603131/)

smessmer committed:
Update on "Register ATen ops with c10"


Reasons for some ops to be excluded:
- `Tensor?(a!)` (i.e. optional tensor with annotations) not supported in c++ function schema parser yet
- `-> void` in native_functions.yaml vs `-> ()` expected by function schema parser
- out functions have different argument order in C++ as in the jit schema
- `Tensor?` (i.e. optional tensor) doesn't work nicely with undefined tensor sometimes being undefined tensor and sometimes being None.
- fixed-size arrays like `int[3]` not supported in c10 yet

These will be fixed in separate diffs and then the exclusion tag will be removed.

Differential Revision: [D16603131](https://our.internmc.facebook.com/intern/diff/D16603131/)

ezyang commented:
Can there be some high level description of how this diff is put together? Thanks!

smessmer committed:
Update on "Register ATen ops with c10"


Reasons for some ops to be excluded:
- `Tensor?(a!)` (i.e. optional tensor with annotations) not supported in c++ function schema parser yet
- `-> void` in native_functions.yaml vs `-> ()` expected by function schema parser
- out functions have different argument order in C++ as in the jit schema
- `Tensor?` (i.e. optional tensor) doesn't work nicely with undefined tensor sometimes being undefined tensor and sometimes being None.
- fixed-size arrays like `int[3]` not supported in c10 yet

These will be fixed in separate diffs and then the exclusion tag will be removed.

Differential Revision: [D16603131](https://our.internmc.facebook.com/intern/diff/D16603131/)

smessmer commented:
- For each operator with `use_c10_dispatcher: True`, additionally generate a c10 registration line in `TypeDefault.cpp`, `CPUType.cpp`, and other backend files.
- This doesn't change `globalATenDispatch` yet, the c10 registration is purely additional and the operator calling path doesn't change. A diff further up the stack will change these things.
- Enable the `use_c10_dispatcher: True` flag for about ~70% of operators
- This also changes the c10->jit operator export because ATen ops are already exported to JIT directly and we don't want to export the registered c10 ops because they would clash
- For this, we need a way to recognize if a certain operator is already moved from ATen to c10, this is done by generating a `OpsAlreadyMovedToC10.cpp` file with the list. A diff further up in the stack will also need this file to make sure we don't break the backend extension API for these ops.

smessmer committed:
Update on "Register ATen ops with c10"


Changes in this PR:
- For each operator with use_c10_dispatcher: True, additionally generate a c10 registration line in TypeDefault.cpp, CPUType.cpp, and other backend files.
- This doesn't change globalATenDispatch yet, the c10 registration is purely additional and the operator calling path doesn't change. A diff further up the stack will change these things.
- Enable the use_c10_dispatcher: True flag for about ~70% of operators
This also changes the c10->jit operator export because ATen ops are already exported to JIT directly and we don't want to export the registered c10 ops because they would clash
- For this, we need a way to recognize if a certain operator is already moved from ATen to c10, this is done by generating a OpsAlreadyMovedToC10.cpp file with the list. A diff further up in the stack will also need this file to make sure we don't break the backend extension API for these ops.

Reasons for some ops to be excluded (i.e. not have the use_c10_dispatcher flag set to true):
- Tensor?(a!) (i.e. optional tensor with annotations) not supported in c++ function schema parser yet
- -> void in native_functions.yaml vs -> () expected by function schema parser
- out functions have different argument order in C++ as in the jit schema
Tensor? (i.e. optional tensor) doesn't work nicely with undefined tensor sometimes being undefined tensor and sometimes being None.
- fixed-size arrays like int[3] not supported in c10 yet
These will be fixed in separate diffs and then the exclusion tag will be removed.

Differential Revision: [D16603131](https://our.internmc.facebook.com/intern/diff/D16603131/)

smessmer committed:
Update on "Register ATen ops with c10"


Changes in this PR:
- For each operator with use_c10_dispatcher: True, additionally generate a c10 registration line in TypeDefault.cpp, CPUType.cpp, and other backend files.
- This doesn't change globalATenDispatch yet, the c10 registration is purely additional and the operator calling path doesn't change. A diff further up the stack will change these things.
- Enable the use_c10_dispatcher: True flag for about ~70% of operators
This also changes the c10->jit operator export because ATen ops are already exported to JIT directly and we don't want to export the registered c10 ops because they would clash
- For this, we need a way to recognize if a certain operator is already moved from ATen to c10, this is done by generating a OpsAlreadyMovedToC10.cpp file with the list. A diff further up in the stack will also need this file to make sure we don't break the backend extension API for these ops.

Reasons for some ops to be excluded (i.e. not have the use_c10_dispatcher flag set to true):
- Tensor?(a!) (i.e. optional tensor with annotations) not supported in c++ function schema parser yet
- -> void in native_functions.yaml vs -> () expected by function schema parser
- out functions have different argument order in C++ as in the jit schema
Tensor? (i.e. optional tensor) doesn't work nicely with undefined tensor sometimes being undefined tensor and sometimes being None.
- fixed-size arrays like int[3] not supported in c10 yet
These will be fixed in separate diffs and then the exclusion tag will be removed.

Differential Revision: [D16603131](https://our.internmc.facebook.com/intern/diff/D16603131/)

smessmer committed:
Update on "Register ATen ops with c10"


Changes in this PR:
- For each operator with use_c10_dispatcher: True, additionally generate a c10 registration line in TypeDefault.cpp, CPUType.cpp, and other backend files.
- This doesn't change globalATenDispatch yet, the c10 registration is purely additional and the operator calling path doesn't change. A diff further up the stack will change these things.
- Enable the use_c10_dispatcher: True flag for about ~70% of operators
This also changes the c10->jit operator export because ATen ops are already exported to JIT directly and we don't want to export the registered c10 ops because they would clash
- For this, we need a way to recognize if a certain operator is already moved from ATen to c10, this is done by generating a OpsAlreadyMovedToC10.cpp file with the list. A diff further up in the stack will also need this file to make sure we don't break the backend extension API for these ops.

Reasons for some ops to be excluded (i.e. not have the use_c10_dispatcher flag set to true):
- Tensor?(a!) (i.e. optional tensor with annotations) not supported in c++ function schema parser yet
- -> void in native_functions.yaml vs -> () expected by function schema parser
- out functions have different argument order in C++ as in the jit schema
Tensor? (i.e. optional tensor) doesn't work nicely with undefined tensor sometimes being undefined tensor and sometimes being None.
- fixed-size arrays like int[3] not supported in c10 yet
These will be fixed in separate diffs and then the exclusion tag will be removed.

Differential Revision: [D16603131](https://our.internmc.facebook.com/intern/diff/D16603131/)

smessmer committed:
Update on "Register ATen ops with c10"


Changes in this PR:
- For each operator with use_c10_dispatcher: True, additionally generate a c10 registration line in TypeDefault.cpp, CPUType.cpp, and other backend files.
- This doesn't change globalATenDispatch yet, the c10 registration is purely additional and the operator calling path doesn't change. A diff further up the stack will change these things.
- Enable the use_c10_dispatcher: True flag for about ~70% of operators
This also changes the c10->jit operator export because ATen ops are already exported to JIT directly and we don't want to export the registered c10 ops because they would clash
- For this, we need a way to recognize if a certain operator is already moved from ATen to c10, this is done by generating a OpsAlreadyMovedToC10.cpp file with the list. A diff further up in the stack will also need this file to make sure we don't break the backend extension API for these ops.

Reasons for some ops to be excluded (i.e. not have the use_c10_dispatcher flag set to true):
- Tensor?(a!) (i.e. optional tensor with annotations) not supported in c++ function schema parser yet
- -> void in native_functions.yaml vs -> () expected by function schema parser
- out functions have different argument order in C++ as in the jit schema
Tensor? (i.e. optional tensor) doesn't work nicely with undefined tensor sometimes being undefined tensor and sometimes being None.
- fixed-size arrays like int[3] not supported in c10 yet
These will be fixed in separate diffs and then the exclusion tag will be removed.

Differential Revision: [D16603131](https://our.internmc.facebook.com/intern/diff/D16603131/)

ezyang commented on aten/src/ATen/templates/OpsAlreadyMovedToC10.cpp:
```diff
@@ -0,0 +1,15 @@
+#include "OpsAlreadyMovedToC10.h"
```
This template needs 

```
// ${generated_comment}
```

ezyang replied:
nit: `#include <ATen/core/OpsAlreadyMovedToC10.h>`. This is inline with prevailing fbcode style and also the choice of quotes versus angled brackets sometimes matters on MSVC.

ezyang commented on aten/src/ATen/core/OpsAlreadyMovedToC10.cpp:
```diff
@@ -0,0 +1,598 @@
+
+// TODO Once all ATen ops are moved to c10, this file should be removed
+
+const std::unordered_set<c10::OperatorName>& aten_ops_already_moved_to_c10() {
```
IIUC, @gchanan is trying to get rid of the checked in autogenerated files. So I'm pretty unhappy that we're adding another checked in autogenerated file in this patch. Is it *absolutely* necessary that this file lives in core? In particular, I find it hard to believe that mobile actually needs to depend on this file (and it really shouldn't, because you're going to pay for binary size to hold the unordered set and strings). If mobile doesn't depend on this file, there is no reason for the autogenerated file to be checked in.

smessmer replied:
I've tried several other alternatives like adding this code to VariableType.cpp or TypeDefault.cpp before but they didn't work. ATenDispatch.h/cpp depends on this and is part of the ATen/core build and not having this file in ATen/core caused linker errors.

I don't really like this either but I think it's ok as an intermediate solution.

ezyang replied:
My comment at https://github.com/pytorch/pytorch/pull/23668#issuecomment-530026140 is applicable here

ezyang commented on aten/src/ATen/core/op_registration/op_registration.h:
```diff
@@ -317,7 +319,10 @@ class CAFFE2_API RegisterOperators final {
+        // setting cache creator to nullptr so calling the kernel doesn't need to call it, which would be expensive
+        // This, however, only works if there are no constructor parameters (i.e. no runtime function pointer)
+        // Backend extensions use runtime function pointers, so we need a cache if sizeof...(ConstructorParameters) != 0
+        (sizeof...(ConstructorParameters) == 0) ? KernelCacheCreatorFunction(nullptr) : detail::KernelFactory<KernelFunctor, guts::decay_t<ConstructorParameters>...>(std::forward<ConstructorParameters>(constructorParameters)...),
```
So, this patch also adds support for some operators which didn't work before?

smessmer replied:
This is not for new operators but for backend extensions. All ATen operators in native_functions.yaml codegen registrations that don't need this, but if you look at "Additional concern: Extension backends" in https://github.com/pytorch/pytorch/issues/24132, you see the plan to keep globalATenDispatch().registerOp() working until we moved all ops to c10, and only then break all ops at once for them. To keep globalATenDispatch().registerOp() working, we need this.

ezyang commented on aten/src/ATen/function_wrapper.py:
```diff
@@ -119,6 +119,18 @@ def TypedDict(name, attrs, total=True):  # type: ignore
+C10_BACKEND_FUNCTION_REGISTRATION = CodeTemplate("""\
+.op("${schema_string}", torch::RegisterOperators::options()
+  .impl_unboxedOnlyKernel<${return_type} (${formals_types}), &${Type}::${api_name}>(
+    c10::backendToTensorTypeId(Backend::${Backend}))
```
TBH, you should just codegen the TensorTypeId directly here...

ezyang commented on aten/src/ATen/function_wrapper.py:
```diff
@@ -119,6 +119,18 @@ def TypedDict(name, attrs, total=True):  # type: ignore
+.op("${schema_string}", torch::RegisterOperators::options()
+  .impl_unboxedOnlyKernel<${return_type} (${formals_types}), &${Type}::${api_name}>(
+    c10::backendToTensorTypeId(Backend::${Backend}))
+  .aliasAnalysis(c10::AliasAnalysisKind::FROM_SCHEMA))
```
Why isn't this alias analysis setting the default?

smessmer replied:
Because we want the custom op API to be conservative by default.

ezyang commented:
It looks like the patch should work. My primary objection is procedural: adding another checked in autogenerated file.

smessmer committed:
Update on "Register ATen ops with c10"


Changes in this PR:
- For each operator with use_c10_dispatcher: True, additionally generate a c10 registration line in TypeDefault.cpp, CPUType.cpp, and other backend files.
- This doesn't change globalATenDispatch yet, the c10 registration is purely additional and the operator calling path doesn't change. A diff further up the stack will change these things.
- Enable the use_c10_dispatcher: True flag for about ~70% of operators
This also changes the c10->jit operator export because ATen ops are already exported to JIT directly and we don't want to export the registered c10 ops because they would clash
- For this, we need a way to recognize if a certain operator is already moved from ATen to c10, this is done by generating a OpsAlreadyMovedToC10.cpp file with the list. A diff further up in the stack will also need this file to make sure we don't break the backend extension API for these ops.

Reasons for some ops to be excluded (i.e. not have the use_c10_dispatcher flag set to true):
- Tensor?(a!) (i.e. optional tensor with annotations) not supported in c++ function schema parser yet
- -> void in native_functions.yaml vs -> () expected by function schema parser
- out functions have different argument order in C++ as in the jit schema
Tensor? (i.e. optional tensor) doesn't work nicely with undefined tensor sometimes being undefined tensor and sometimes being None.
- fixed-size arrays like int[3] not supported in c10 yet
These will be fixed in separate diffs and then the exclusion tag will be removed.

Differential Revision: [D16603131](https://our.internmc.facebook.com/intern/diff/D16603131/)

smessmer committed:
Update on "Register ATen ops with c10"


Changes in this PR:
- For each operator with use_c10_dispatcher: True, additionally generate a c10 registration line in TypeDefault.cpp, CPUType.cpp, and other backend files.
- This doesn't change globalATenDispatch yet, the c10 registration is purely additional and the operator calling path doesn't change. A diff further up the stack will change these things.
- Enable the use_c10_dispatcher: True flag for about ~70% of operators
This also changes the c10->jit operator export because ATen ops are already exported to JIT directly and we don't want to export the registered c10 ops because they would clash
- For this, we need a way to recognize if a certain operator is already moved from ATen to c10, this is done by generating a OpsAlreadyMovedToC10.cpp file with the list. A diff further up in the stack will also need this file to make sure we don't break the backend extension API for these ops.

Reasons for some ops to be excluded (i.e. not have the use_c10_dispatcher flag set to true):
- Tensor?(a!) (i.e. optional tensor with annotations) not supported in c++ function schema parser yet
- -> void in native_functions.yaml vs -> () expected by function schema parser
- out functions have different argument order in C++ as in the jit schema
Tensor? (i.e. optional tensor) doesn't work nicely with undefined tensor sometimes being undefined tensor and sometimes being None.
- fixed-size arrays like int[3] not supported in c10 yet
These will be fixed in separate diffs and then the exclusion tag will be removed.

Differential Revision: [D16603131](https://our.internmc.facebook.com/intern/diff/D16603131/)

smessmer committed:
Update on "Register ATen ops with c10"


Changes in this PR:
- For each operator with use_c10_dispatcher: True, additionally generate a c10 registration line in TypeDefault.cpp, CPUType.cpp, and other backend files.
- This doesn't change globalATenDispatch yet, the c10 registration is purely additional and the operator calling path doesn't change. A diff further up the stack will change these things.
- Enable the use_c10_dispatcher: True flag for about ~70% of operators
This also changes the c10->jit operator export because ATen ops are already exported to JIT directly and we don't want to export the registered c10 ops because they would clash
- For this, we need a way to recognize if a certain operator is already moved from ATen to c10, this is done by generating a OpsAlreadyMovedToC10.cpp file with the list. A diff further up in the stack will also need this file to make sure we don't break the backend extension API for these ops.

Reasons for some ops to be excluded (i.e. not have the use_c10_dispatcher flag set to true):
- Tensor?(a!) (i.e. optional tensor with annotations) not supported in c++ function schema parser yet
- -> void in native_functions.yaml vs -> () expected by function schema parser
- out functions have different argument order in C++ as in the jit schema
Tensor? (i.e. optional tensor) doesn't work nicely with undefined tensor sometimes being undefined tensor and sometimes being None.
- fixed-size arrays like int[3] not supported in c10 yet
These will be fixed in separate diffs and then the exclusion tag will be removed.

Differential Revision: [D16603131](https://our.internmc.facebook.com/intern/diff/D16603131/)

smessmer committed:
Update on "Register ATen ops with c10"


Changes in this PR:
- For each operator with use_c10_dispatcher: True, additionally generate a c10 registration line in TypeDefault.cpp, CPUType.cpp, and other backend files.
- This doesn't change globalATenDispatch yet, the c10 registration is purely additional and the operator calling path doesn't change. A diff further up the stack will change these things.
- Enable the use_c10_dispatcher: True flag for about ~70% of operators
This also changes the c10->jit operator export because ATen ops are already exported to JIT directly and we don't want to export the registered c10 ops because they would clash
- For this, we need a way to recognize if a certain operator is already moved from ATen to c10, this is done by generating a OpsAlreadyMovedToC10.cpp file with the list. A diff further up in the stack will also need this file to make sure we don't break the backend extension API for these ops.

Reasons for some ops to be excluded (i.e. not have the use_c10_dispatcher flag set to true):
- Tensor?(a!) (i.e. optional tensor with annotations) not supported in c++ function schema parser yet
- -> void in native_functions.yaml vs -> () expected by function schema parser
- out functions have different argument order in C++ as in the jit schema
Tensor? (i.e. optional tensor) doesn't work nicely with undefined tensor sometimes being undefined tensor and sometimes being None.
- fixed-size arrays like int[3] not supported in c10 yet
These will be fixed in separate diffs and then the exclusion tag will be removed.

Differential Revision: [D16603131](https://our.internmc.facebook.com/intern/diff/D16603131/)

smessmer committed:
Update on "Register ATen ops with c10"


Changes in this PR:
- For each operator with use_c10_dispatcher: True, additionally generate a c10 registration line in TypeDefault.cpp, CPUType.cpp, and other backend files.
- This doesn't change globalATenDispatch yet, the c10 registration is purely additional and the operator calling path doesn't change. A diff further up the stack will change these things.
- Enable the use_c10_dispatcher: True flag for about ~70% of operators
This also changes the c10->jit operator export because ATen ops are already exported to JIT directly and we don't want to export the registered c10 ops because they would clash
- For this, we need a way to recognize if a certain operator is already moved from ATen to c10, this is done by generating a OpsAlreadyMovedToC10.cpp file with the list. A diff further up in the stack will also need this file to make sure we don't break the backend extension API for these ops.

Reasons for some ops to be excluded (i.e. not have the use_c10_dispatcher flag set to true):
- Tensor?(a!) (i.e. optional tensor with annotations) not supported in c++ function schema parser yet
- -> void in native_functions.yaml vs -> () expected by function schema parser
- out functions have different argument order in C++ as in the jit schema
Tensor? (i.e. optional tensor) doesn't work nicely with undefined tensor sometimes being undefined tensor and sometimes being None.
- fixed-size arrays like int[3] not supported in c10 yet
These will be fixed in separate diffs and then the exclusion tag will be removed.

Differential Revision: [D16603131](https://our.internmc.facebook.com/intern/diff/D16603131/)

smessmer committed:
Update on "Register ATen ops with c10"


Changes in this PR:
- For each operator with use_c10_dispatcher: True, additionally generate a c10 registration line in TypeDefault.cpp, CPUType.cpp, and other backend files.
- This doesn't change globalATenDispatch yet, the c10 registration is purely additional and the operator calling path doesn't change. A diff further up the stack will change these things.
- Enable the use_c10_dispatcher: True flag for about ~70% of operators
This also changes the c10->jit operator export because ATen ops are already exported to JIT directly and we don't want to export the registered c10 ops because they would clash
- For this, we need a way to recognize if a certain operator is already moved from ATen to c10, this is done by generating a OpsAlreadyMovedToC10.cpp file with the list. A diff further up in the stack will also need this file to make sure we don't break the backend extension API for these ops.

Reasons for some ops to be excluded (i.e. not have the use_c10_dispatcher flag set to true):
- Tensor?(a!) (i.e. optional tensor with annotations) not supported in c++ function schema parser yet
- -> void in native_functions.yaml vs -> () expected by function schema parser
- out functions have different argument order in C++ as in the jit schema
Tensor? (i.e. optional tensor) doesn't work nicely with undefined tensor sometimes being undefined tensor and sometimes being None.
- fixed-size arrays like int[3] not supported in c10 yet
These will be fixed in separate diffs and then the exclusion tag will be removed.

Differential Revision: [D16603131](https://our.internmc.facebook.com/intern/diff/D16603131/)

ezyang reviewed (approved):
Reiterating from the closed PR: I think the RegisterOperators is bad (having to specify schema string twice is a big no no) and needs to be fixed.

smessmer committed:
Update on "Register ATen ops with c10"


Changes in this PR:
- For each operator with use_c10_dispatcher: True, additionally generate a c10 registration line in TypeDefault.cpp, CPUType.cpp, and other backend files.
- This doesn't change globalATenDispatch yet, the c10 registration is purely additional and the operator calling path doesn't change. A diff further up the stack will change these things.
- Enable the use_c10_dispatcher: True flag for about ~70% of operators
This also changes the c10->jit operator export because ATen ops are already exported to JIT directly and we don't want to export the registered c10 ops because they would clash
- For this, we need a way to recognize if a certain operator is already moved from ATen to c10, this is done by generating a OpsAlreadyMovedToC10.cpp file with the list. A diff further up in the stack will also need this file to make sure we don't break the backend extension API for these ops.

Reasons for some ops to be excluded (i.e. not have the use_c10_dispatcher flag set to true):
- Tensor?(a!) (i.e. optional tensor with annotations) not supported in c++ function schema parser yet
- -> void in native_functions.yaml vs -> () expected by function schema parser
- out functions have different argument order in C++ as in the jit schema
Tensor? (i.e. optional tensor) doesn't work nicely with undefined tensor sometimes being undefined tensor and sometimes being None.
- fixed-size arrays like int[3] not supported in c10 yet
These will be fixed in separate diffs and then the exclusion tag will be removed.

Differential Revision: [D16603131](https://our.internmc.facebook.com/intern/diff/D16603131/)
